### PR TITLE
Patch: Compress transaction history

### PR DESCRIPTION
**Pull request by Gudahtt**
## **Description**

Compress transaction history down to 100 entries and truncate any existing transaction historys to just 100 entries. This should fix reports we've seen from production of extremely poor performance and crashes caused by long transaction histories, and will also prevent this problem from happening again.

Transaction history has been truncated because in the extreme cases, it would be prohibitively expensive to compress. The downside is that some state of how the transaction has changed may be lost. But this is unlikely to impact users because we only show a limited number of events from the transaction history in our UI, and these events are more likely to be at the beginning of long transaction histories. Even if a displayed event is lost, the impact on the UI is minimal (it's only shown on the transaction details page under "Activity log", and only for informational purposes).

For details on how the transaction compression works, and how it prevents history size from growing unbouned, see https://github.com/MetaMask/core/pull/4555

The transaction controller change has been applied using a patch. The patch was generated from the core repository branch `patch/transaction-controller-v32-compress-history`. It will be made on `develop` at a later date because it is blocked by other controller updates at the moment.

The truncation is performed by a migration that was added in #26291 and cherry-picked into this branch

[Open in GitHub Codespaces](https://codespaces.new/MetaMask/metamask-extension/pull/26236?quickstart=1)

## **Related issues**

Relates to #9372

## **Manual testing steps**

To test the migration:
* Checkout v11.15.6 and create a dev build (unfortunately the repro steps don't work on MV3 due to snow being enabled even in dev builds, so we're using a pre-MV3 build)
* Install the dev build from the `dist/chrome` directory and proceed through onboarding
* Make a single transaction
  * Any chain, any transaction, as long as it's approved. We don't even need to wait for it to settle.
* Run this command in the background console to add a huge history to the transaction controller and restart the extension:
  ```
  chrome.storage.local.get(
    null,
    (state) => {
      state.data.TransactionController.transactions[0].history.push(
        ...[...Array(100000).keys()].map(() => [
          {
            value: '[ethjs-rpc] rpc error with payload {"id":[number],"jsonrpc":"2.0","params":["0x"],"method":"eth_getTransactionReceipt"} { "code": -32603, "message": "Internal JSON-RPC error.", "data": { "code": -32602, "message": "invalid argument 0: hex string has length 0, want 64 for common.Hash", "cause": null }, "stack": ...',
            path: '/warning/error',
            op: 'replace'
          },
          {
            note: 'transactions/pending-tx-tracker#event: tx:warning',
            value: 'There was a problem loading this transaction.',
            path: '/warning/message',
            op: 'replace'
          },
        ]),
      );
      chrome.storage.local.set(state, () => chrome.runtime.reload());
    }
  );
  ```
  * The extension should become extremely slow
* Disable the extension
* Switch to this branch and create a dev build
* Enable and reload the extension
  * The extension should no longer be slow

To test that the compression is working, we would want to get a transaction in a "stuck pending" state where an error is captured each iteration. I am not yet sure how to do that unfortunately.

## **Screenshots/Recordings**

N/A

## **Pre-merge author checklist**

- [x] I've followed [MetaMask Contributor Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Extension Coding Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've completed the PR template to the best of my ability
- [x] I’ve included tests if applicable
- [x] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [x] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

### CODE REVIEW

File: app/scripts/migrations/120.2.ts [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+    return state;
+  } else if (!Array.isArray(transactionControllerState.transactions)) {
+    global.sentry?.captureException(
+      `Migration ${version}: Invalid TransactionController transactions state of type '${typeof transactionControllerState.transactions}'`,
+    );
+    return state;
+  }
+
+  const validTransactions =
+    transactionControllerState.transactions.filter(isObject);
```
Comment (DDDDDanica) [inline app/scripts/migrations/120.2.ts:60]: Question: under which circumstance transactions is not an object?

File: app/scripts/migrations/120.2.ts [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+      (transaction) => !hasValidTransactionHistory(transaction),
+    );
+    global.sentry?.captureException(
+      `Migration ${version}: Invalid transaction history of type '${typeof invalidTransaction?.history}'`,
+    );
+    return state;
+  }
+
+  for (const transaction of validHistoryTransactions) {
+    if (transaction.history && transaction.history.length > 100) {
```
Comment (DDDDDanica) [inline app/scripts/migrations/120.2.ts:87]: Could we apply same constant to the limit here, as `const MAX_TRANSACTION_HISTORY_LENGTH = 100;`

**Comment by danjm** [comment]
Why is there a 120.2 and 120.3 is this PR?

File: app/scripts/migrations/120.2.ts (hunk 1 above)
Comment (Gudahtt) [reply app/scripts/migrations/120.2.ts:60]: They should always be objects. This is meant to validate that the state is not corrupted somehow. We validate everything that the migration assumes, and in this case we are assuming it's an object so we validate that first.

File: app/scripts/migrations/120.2.ts (hunk 2 above)
Comment (Gudahtt) [reply app/scripts/migrations/120.2.ts:87]: Good catch, though as Dan pointed out this entire file is a duplicate, so I will remove it entirely

**Comment by Gudahtt** [comment]
>There is both a 120.2 and 120.3 is this PR. The code is almost identical between the two... I assume this is not intentional?
Correct, that was a mistake. 120.2 was the old copy, 120.3 was cherry-picked from `develop`. The old one has been deleted in 497d64f57d

I'm leaving a gap here for 120.2 because we still need to cherry-pick that in as well

**Comment by danjm** [comment]
I have successfully manually tested this. If the transaction history reaches length 100, then further additions to the history do not increase the length, but the latest additions are correctly added (whether those are errors or otherwise)

To test, I made 100 edits to gas price, and then confirmed the transaction. I also did the same, but with some code in the transaction controller edited, such that the next resubmission after first submission would fail.

**Review by danjm — APPROVED** [review]

**Pull request merged**